File: internal/status/status.go
package status

import (
	"sort"

	"github.com/heventure/hermes-agent-cluster/internal/cluster"
	"github.com/heventure/hermes-agent-cluster/internal/lease"
	"github.com/heventure/hermes-agent-cluster/internal/scheduler"
)

// StatusEntry represents a single task in the global status view.
type StatusEntry struct {
	TaskID      string              `json:"task_id"`
	TaskTitle   string              `json:"task_title"`
	TaskStatus  scheduler.TaskStatus `json:"task_status"`
	NodeID      string              `json:"node_id"`
	NodeName    string              `json:"node_name"`
	NodeStatus  cluster.NodeStatus  `json:"node_status"`
	Capabilities []string           `json:"capabilities,omitempty"`
	Requires    []string            `json:"requires,omitempty"`
	LeaseStatus string              `json:"lease_status,omitempty"`
	FailReason  string              `json:"fail_reason,omitempty"`
}

// Summary provides aggregate counts for the global status view.
type Summary struct {
	TotalNodes   int                          `json:"total_nodes"`
	OnlineNodes  int                          `json:"online_nodes"`
	TotalTasks   int                          `json:"total_tasks"`
	TasksByStatus map[string]int              `json:"tasks_by_status"`
	ActiveLeases int                          `json:"active_leases"`
}

// StatusView aggregates all cluster state into a queryable view.
type StatusView struct {
	Registry  *cluster.Registry
	TaskStore *scheduler.TaskStore
	LeaseMgr  *lease.Manager
}

// NewStatusView creates a new StatusView.
func NewStatusView(registry *cluster.Registry, taskStore *scheduler.TaskStore, leaseMgr *lease.Manager) *StatusView {
	return &StatusView{
		Registry:  registry,
		TaskStore: taskStore,
		LeaseMgr:  leaseMgr,
	}
}

// Filter holds optional query parameters for filtering the status view.
type Filter struct {
	NodeID     string // filter by node ID
	Status     string // filter by task status
	Capability string // filter by capability (any node with this cap)
}

// Query returns the filtered status view and summary.
func (sv *StatusView) Query(f Filter) ([]StatusEntry, Summary) {
	tasks := sv.TaskStore.GetAll()
	nodes := sv.Registry.GetAll()
	activeLeases := sv.LeaseMgr.GetActiveLeases()

	// Build lookup maps
	nodeMap := make(map[string]*cluster.Node, len(nodes))
	for _, n := range nodes {
		nodeMap[n.ID] = n
	}
	leaseByTask := make(map[string]*lease.Lease, len(activeLeases))
	for _, l := range activeLeases {
		leaseByTask[l.TaskID] = l
	}

	// Build summary (unfiltered)
	summary := Summary{
		TotalNodes:   len(nodes),
		OnlineNodes:  sv.Registry.OnlineCount(),
		TotalTasks:   len(tasks),
		TasksByStatus: make(map[string]int),
		ActiveLeases: len(activeLeases),
	}
	for _, t := range tasks {
		summary.TasksByStatus[string(t.Status)]++
	}

	// Build entries with filtering
	var entries []StatusEntry
	for _, t := range tasks {
		// Filter by task status
		if f.Status != "" && string(t.Status) != f.Status {
			continue
		}

		entry := StatusEntry{
			TaskID:     t.ID,
			TaskTitle:  t.Title,
			TaskStatus: t.Status,
			Requires:   t.Requires,
			FailReason: t.FailReason,
		}

		// Resolve node info
		if t.AssignedTo != "" {
			if node, ok := nodeMap[t.AssignedTo]; ok {
				entry.NodeID = node.ID
				entry.NodeName = node.Name
				entry.NodeStatus = node.Status
				entry.Capabilities = node.Capabilities
			} else {
				entry.NodeID = t.AssignedTo
			}
		}

		// Attach lease info
		if l, ok := leaseByTask[t.ID]; ok {
			entry.LeaseStatus = string(l.Status)
		}

		// Filter by node ID
		if f.NodeID != "" && entry.NodeID != f.NodeID {
			continue
		}

		// Filter by capability
		if f.Capability != "" {
			hasCap := false
			for _, c := range entry.Capabilities {
				if c == f.Capability {
					hasCap = true
					break
				}
			}
			if !hasCap {
				continue
			}
		}

		entries = append(entries, entry)
	}

	// Sort by task status then task ID for stable output
	sort.Slice(entries, func(i, j int) bool {
		if entries[i].TaskStatus != entries[j].TaskStatus {
			return entries[i].TaskStatus < entries[j].TaskStatus
		}
		return entries[i].TaskID < entries[j].TaskID
	})

	return entries, summary
}

File: hermes_cluster/routers/status.py
"""Status / Summary endpoints — /api/v1/status, /api/v1/summary"""

from fastapi import APIRouter, Query

from ..state import ClusterState

router = APIRouter(prefix="/api/v1", tags=["status"])

_state: ClusterState = None


def init(state: ClusterState):
    global _state
    _state = state


@router.get("/status")
async def status(
    node: str = Query("", description="Filter by node ID"),
    status: str = Query("", description="Filter by status"),
    capability: str = Query("", description="Filter by capability"),
):
    tasks = _state.get_all_tasks()
    nodes = _state.get_all_nodes()

    # Build entries matching Go's status view
    entries = []
    for task in tasks:
        entry = {
            "task_id": task.id,
            "title": task.title,
            "status": task.status.value if hasattr(task.status, 'value') else task.status,
            "priority": task.priority,
            "assigned_to": task.assigned_to,
            "requires": task.requires,
        }
        # Filter
        if node and task.assigned_to != node:
            continue
        if status and entry["status"] != status:
            continue
        if capability and capability not in task.requires:
            continue
        entries.append(entry)

    # Build summary
    task_counts = _state.task_counts()
    summary = {
        "total_tasks": task_counts["total"],
        "pending": task_counts.get("pending", 0),
        "ready": task_counts.get("ready", 0),
        "running": task_counts.get("running", 0),
        "completed": task_counts.get("completed", 0),
        "failed": task_counts.get("failed", 0),
        "total_nodes": _state.node_count(),
        "online_nodes": _state.online_count(),
    }

    return {"entries": entries, "summary": summary}


@router.get("/summary")
async def summary():
    return _state.get_summary()
